MCH: added error handler in raw decoder

aferrero2707 opened the pull request:
The raw decoder handlers are now grouped in a structure,
so that new handlers can be easily added in future if needed.

A backward-compatible version of the createPageDecoder
has been introduced, to be removed once all code using the
O2 decoder will be adapted to use the RawDecoderHandlers structure.

Currently handlers for SAMPA data packets and for decoder errors
are implemented in the code.

The DataDecoder algorithm is also moved into a separate source
file, for better readability and easier unit testing.

Some initial unit tests for the error handling in the User Logic decoder
have been added.

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/include/MCHRawDecoder/DataDecoder.h:
```diff
@@ -0,0 +1,82 @@
+/// \file    DataDecoder.h
+/// \author  Andrea Ferrero
+///
+/// \brief Definition of a data processor to run the raw decoding
```
This DataDecoder is no longer a data processor (in the DPL sense), right ?

aferrero2707 replied:
I have changed the documentation line to
```
Definition of the decoder for the MCH data
```

aphecetche commented on Detectors/MUON/MCH/Raw/Common/src/SampaCluster.cxx:
```diff
@@ -72,6 +72,20 @@ uint16_t SampaCluster::nof10BitWords() const
+  if (isClusterSum()) {
+    tot = chargeSum;
+  } else {
+    for (auto& s : samples) {
```
that's a micro (or even nano ;-) ) comment/optimization, and one I have no idea if it actually has some influence on the performance at all, but just for the sake of code clarity (intent) `const auto&` would probably be clearer.

aferrero2707 replied:
Done. I also propose the following modification in the `nof10BitWords()` function:

From
```
  if (isClusterSum()) {
    n10 += 2; // 20 bits (chargesum)
  } else {
    for (auto s : samples) {
      ++n10; // 10 bits for each sample
    }
  }
```

to

```
  if (isClusterSum()) {
    n10 += 2; // 20 bits (chargesum)
  } else {
    n10 += samples.size();
  }
```

aphecetche replied:
yep, sure.

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/include/MCHRawDecoder/DataDecoder.h:
```diff
@@ -0,0 +1,82 @@
+
+// custom hash for OrbitInfo objects
+struct OrbitInfoHash {
+  std::size_t operator()(OrbitInfo const& info) const noexcept
```
Our coding conventions recommend putting the `const` before the type, not after 
 https://rawgit.com/AliceO2Group/CodingGuidelines/master/naming_formatting.html?showone=Where_to_put_const#Where_to_put_const

aferrero2707 replied:
Done

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/include/MCHRawDecoder/PageDecoder.h:
```diff
@@ -45,6 +45,15 @@ PageDecoder createPageDecoder(RawBuffer rdhBuffer,
+                              FeeLink2SolarMapper fee2solar);
+
+// Alternative versions of the same functions, taking a SampaChannelHandler as parameter.
+// For backward compatibility only, to be removed in future.
```
Instead of the backward compat. warning I would use the C++ attribute : 
```
[[deprecated("Use createPageDecoder(RawBuffer,RawDecoderHandlers) instead.")]]
```

aphecetche replied:
or simply drop it right now ? I don't see it used in O2 at all with this PR (or is it used in QC maybe ?)

aferrero2707 replied:
I have put the deprecated attribute. The old function is used in QC, so we need to keep it a round for a short while until QC is modified as well.

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/include/MCHRawDecoder/RawDecoderHandlers.h:
```diff
@@ -27,6 +27,18 @@ namespace raw
 using SampaChannelHandler = std::function<void(DsElecId dsId,
                                                uint8_t channel,
                                                SampaCluster)>;
+
```
note that the include guards no longer reflects the file name

aferrero2707 replied:
Fixed

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/include/MCHRawDecoder/RawDecoderHandlers.h:
```diff
@@ -27,6 +27,18 @@ namespace raw
+                                             int8_t chip,
+                                             uint32_t error)>;
+
+struct RawDecoderHandlers {
```
you probably know by now that I'm picky about names, so here I go again ;-)  
Those are not handling decoder but decoded data really, so instead of `o2::mch::raw::RawDecoderHandlers` (not the stutter btw) I would propose something like `o2::mch::raw::DecodedDataHandlers` or even simply `o2::mch::raw::DataHandlers`  (but that latter one might be too generic)

(I'm also uneasy with the plural, but don't have a good suggestion here, so I'll just shut up ;-) )

aferrero2707 replied:
No problem.  I have renamed `RawDecoderHandlers` to `DecodedDataHandlers` everywhere.

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,304 @@
+#include "MCHMappingInterface/Segmentation.h"
+
+/*
+#include <random>
```
please remove commented out code

aferrero2707 replied:
Done

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/src/DataDecoder.cxx:
```diff
@@ -0,0 +1,304 @@
+    const Segmentation& segment = segmentation(deId);
+
+    padId = segment.findPadByFEE(dsIddet, int(channel));
+    if (mPrint) {
```
Instead of having this mPrint static constant, why not using e.g. `LOG(DEBUG)` that is by default disabled (unless the code is built with CMAKE_BUILD_TYPE=DEBUG) or any other level `LOG(WARNING)`... that can be controlled from the outside ?

aferrero2707 replied:
Generally I would agree, however at least in this specific case the log message requires building a string from the electronics mapping data, and I would prefer to avoid this extra string creation for normal, non-verbose execution.

In fact, this output is really only needed for debugging, and I am changing the code such that it can be enabled/disabled via a command-line option...

aphecetche replied:
ok

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/src/testUserLogicEndpointDecoder.cxx:
```diff
@@ -35,6 +35,138 @@
+  ((0x3FBBul << 50) & 0xFFFC000000000000ul) + 0x1DEEDFEEDDEEDul  // payload size in the RDH
+};
+
+struct RAWDataHeaderV6 {
```
this should be removed

aferrero2707 replied:
Removed

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/src/testUserLogicEndpointDecoder.cxx:
```diff
@@ -35,6 +35,138 @@
 using namespace o2::mch::raw;
 using o2::header::RAWDataHeaderV4;
```
no longer needed I guess

aferrero2707 replied:
Removed as well

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/src/testUserLogicEndpointDecoder.cxx:
```diff
@@ -139,6 +282,31 @@ std::string testPayloadDecode(DsElecId ds1,
+  auto rdhSize = o2::raw::RDHUtils::getHeaderSize(rdhP);
+  auto payloadSize = o2::raw::RDHUtils::getMemorySize(rdhP) - rdhSize;
+
+  //std::cout << "version " << (int)o2::raw::RDHUtils::getVersion(rdhP) <<
```
please remove commented out code

aferrero2707 replied:
Done

aphecetche commented on Detectors/MUON/MCH/Raw/Decoder/include/MCHRawDecoder/DataDecoder.h:
```diff
@@ -0,0 +1,82 @@
+  void reset();
+  void decodeBuffer(gsl::span<const std::byte> page);
+
+  std::vector<o2::mch::Digit>& getOutputDigits() { return mOutputDigits; }
```
those two `getxxx` should return const references and be const functions, e.g. `const std::vector<o2::mch::Digit>& getOutputDigits() const { return ... }`

aferrero2707 replied:
Done

aferrero2707 committed:
[MCH] added error handler in raw decoder

The raw decoder handlers are now grouped in a structure,
so that new handlers can be easily added in future if needed.

A backward-compatible version of the createPageDecoder
has been introduced, to be removed once all code using the
O2 decoder will be adapted to use the RawDecoderHandlers structure.

Currently handlers for SAMPA data packets and for decoder errors
are implemented in the code.

The DataDecoder algorithm is also moved into a separate source
file, for better readability and easier unit testing.

Some initial unit tests for the error handling in the User Logic decoder
have been added.

aferrero2707 committed:
[MCH] Fixed clang errors

aferrero2707 committed:
[MCH] triggered CI builds

aferrero2707 committed:
[MCH] fixed clang format error

aferrero2707 commented:
@shahor02 @ktf the TravisCI job for this pull request is hanging since quite a lot, preventing it to be accepted... do you know what is going on?

The MacOS jobs were also failing during last week due to linkage errors that are independent of this PR, but that seems to be fixed now.

ktf commented:
@aferrero2707 We are phasing out travis in favour of Github Actions as discussed in WP3. I thought I removed it from everywhere but apparently not. I just disabled it in one more place and hopefully next PRs will not have the pending test shown.

aferrero2707 commented:
@ktf thanks a lot for your answer! Does it mean that we can proceed with the merge for this one?
@aphecetche is it ok to merge and close this PR?

ktf commented:
Yes.
